posts/2021/02/pytorch-tensoriterator-internals-update.md

```diff
@@ -0,0 +1,321 @@
+strides--one for the dimension being reduced and one for all other dimensions.
+Take a look at the following example of a somewhat simplified version of the
+[cumsum
+kernel](https://github.com/pytorch/pytorch/blob/d9e6750759b78c68e7d98b80202c67bea7ba24ec/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L67).
+
+For a 1-D input,
+[torch.cumsum](https://pytorch.org/docs/stable/generated/torch.cumsum.html?highlight=cumsum#torch.cumsum)
+calculates the sum of all elements from the beginning of the vector up to and
+including each position in the input. A 2-D input is treated as a list of
+vectors, and the cumulative sum is calculated for each vector. Higher
+dimensional inputs follow the same logic--everything is just a list of 1-D
+vectors.  So to implement a cumulative sum, we must take into account two
+different strides: the stride between elements in a vector (`result_dim_stride`
+and `self_dim_stride` in the example below) and the stride between each vector
+(`strides[0]` and `strides[1]` in the example below).
+
+```cpp
+// A cumulative sum's output is the same size as the input
+at::Tensor result = at::empty_like(self);
+
+at::TensorIteratorConfig iter_config;
+auto iter = iter_config
+  .check_all_same_dtype(false)
+  .resize_outputs(false)
+  .declare_static_shape(self.sizes(), /*squash_dim=*/dim)
+  .add_output(result)
+  .add_input(self)
+  .build();
+
+// Size of dimension to calculate the cumulative sum across
+int64_t self_dim_size = at::native::ensure_nonempty_size(self, dim);
+
+// These strides indicate number of memory-contiguous elements, not bytes,
+// between each successive element in dimension `dim`.
+auto result_dim_stride = at::native::ensure_nonempty_stride(result, dim);
+auto self_dim_stride = at::native::ensure_nonempty_stride(self, dim);
+
+auto loop = [&](char** data, const int64_t* strides, int64_t n) {
+  // There are `n` individual vectors that span across dimension `dim`, so
+  // `n` is equal to the number of elements in `self` divided by the size of
+  // dimension `dim`.
+
+  // These are the byte strides that separate each vector that spans across
+  // dimension `dim`
+  auto* result_data_bytes = data[0];
+  const auto* self_data_bytes = data[1];
+
+  for (int64_t vector_idx = 0; vector_idx < n; ++vector_idx) {
+
+    // Calculate cumulative sum for each element of the vector
+    auto cumulative_sum = (at::acc_type<float, false>) 0;
+    for (int64_t elem_idx = 0; elem_idx < self_dim_size; ++elem_idx) {
+      const auto* self_data = reinterpret_cast<const float*>(self_data_bytes);
+      auto* result_data = reinterpret_cast<float*>(result_data_bytes);
+      cumulative_sum += self_data[elem_idx * self_dim_stride];
+      result_data[elem_idx * result_dim_stride] = (float)cumulative_sum;
+    }
+
+    // Go to the next vector
+    result_data_bytes += strides[0];
+    self_data_bytes += strides[1];
+  }
+};
+
+iter.for_each(loop);
+```
+
+### Helper functions
```

Comment:
Hey @nikitaved, could you take a look at this new section when you get a chance? Thanks!